cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/AppFabricServer.java

```diff
@@ -110,36 +102,31 @@ public class AppFabricServer extends AbstractIdleService {
-                         @Named(Constants.AppFabric.HANDLERS_BINDING) Set<HttpHandler> handlers,
-                         @Nullable MetricsCollectionService metricsCollectionService,
-                         ProgramRuntimeService programRuntimeService,
-                         RunRecordCorrectorService runRecordCorrectorService,
-                         ProgramRunStatusMonitorService programRunStatusMonitorService,
-                         ApplicationLifecycleService applicationLifecycleService,
-                         ProgramNotificationSubscriberService programNotificationSubscriberService,
-                         ProgramStopSubscriberService programStopSubscriberService,
-                         @Named("appfabric.services.names") Set<String> servicesNames,
-                         @Named("appfabric.handler.hooks") Set<String> handlerHookNames,
-                         AuditLogSubscriberService auditLogSubscriberService,
-                         CoreSchedulerService coreSchedulerService,
-                         CredentialProviderService credentialProviderService,
-                         NamespaceCredentialProviderService namespaceCredentialProviderService,
-                         ProvisioningService provisioningService,
-                         BootstrapService bootstrapService,
-                         SystemAppManagementService systemAppManagementService,
-                         TransactionRunner transactionRunner,
-                         RunRecordMonitorService runRecordCounterService,
-                         CommonNettyHttpServiceFactory commonNettyHttpServiceFactory,
-                         RunDataTimeToLiveService runDataTimeToLiveService,
-                         SourceControlOperationRunner sourceControlOperationRunner,
-                         RepositoryCleanupService repositoryCleanupService,
-                         OperationNotificationSubscriberService operationNotificationSubscriberService) {
+      DiscoveryService discoveryService,
+      @Named(Constants.Service.MASTER_SERVICES_BIND_ADDRESS) InetAddress hostname,
+      @Named(Constants.AppFabric.SERVER_HANDLERS_BINDING) Set<HttpHandler> handlers,
+      @Nullable MetricsCollectionService metricsCollectionService,
+      ProgramRuntimeService programRuntimeService,
+      ProgramRunStatusMonitorService programRunStatusMonitorService,
+      ApplicationLifecycleService applicationLifecycleService,
+      @Named("appfabric.services.names") Set<String> servicesNames,
+      @Named("appfabric.handler.hooks") Set<String> handlerHookNames,
+      CoreSchedulerService coreSchedulerService,
+      CredentialProviderService credentialProviderService,
+      NamespaceCredentialProviderService namespaceCredentialProviderService,
+      ProvisioningService provisioningService,
+      BootstrapService bootstrapService,
+      SystemAppManagementService systemAppManagementService,
```

Comment:
Same here. I would do it once, not N times.

Reply:
SystemAppManagementService runs SystemAppEnableExecutor which depends on ApplicationLifecycleService, ProgramLifecycleService.

Keeping this in Server only and not the Processor.

Reply:
Moved SystemAppManagementService to processor.